chaindataloader/binary.go

```diff
@@ -0,0 +1,227 @@
+package chaindataloader
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+
+	"github.com/btcsuite/btcd/wire"
+	"github.com/lightninglabs/neutrino/headerfs"
+	"github.com/lightningnetwork/lnd/tlv"
+)
+
+/*
+
+================================================================================
+
+ENCODING FORMAT
+================================================================================
+
+Each file has a header of varint size consisting of,
+
+dataType || startHeight || endHeight || chain
+
+in that order.
+*/
+
+// binReader is an internal struct that holds all data the binReader needs
+// to fetch headers.
+type binReader struct {
+	// reader represents the source to be read.
+	reader io.ReadSeeker
+
+	// startHeight represents the height of the first header in the file.
+	startHeight uint32
+
+	// endHeight represents the height of the last header in the file.
+	endHeight uint32
+
+	// offset represents the distance required to read the first header from
+	// the file.
+	initialOffset uint32
+
+	// chain represents the bitcoin network the headers in the file belong to.
+	chain        wire.BitcoinNet
+	dataTypeSize uint32
+}
+
+// newBinaryChainDataLoader initializes a Binary Reader.
+func newBinaryChainDataLoader(reader io.ReadSeeker,
+	typeData dataType) (*binReader, error) {
+
+	// Read data type.
+	dataTypeBytes := [8]byte{}
+	typeOfData, err := tlv.ReadVarInt(reader, &dataTypeBytes)
+	if err != nil {
+		return nil, fmt.Errorf("error obtaining data type "+
+			"of file %w", err)
+	}
+
+	if dataType(typeOfData) != typeData {
+		return nil, fmt.Errorf("data type mismatch: got %v but expected %v",
+			dataType(typeOfData), typeData)
+	}
+
+	var dataTypeSize uint32
+	switch {
+	case typeData == blockHeaders:
+		dataTypeSize = headerfs.BlockHeaderSize
+	default:
+		return nil, fmt.Errorf("got unsupported dataType: %v", typeData)
+	}
+
+	// Read start height of block header file.
+	startBytes := [8]byte{}
+	start, err := tlv.ReadVarInt(reader, &startBytes)
+	if err != nil {
+		return nil, fmt.Errorf("error obtaining start height "+
+			"of file %w", err)
+	}
+
+	// Read end height of block header file.
+	endBytes := [8]byte{}
+	end, err := tlv.ReadVarInt(reader, &endBytes)
+	if err != nil {
+		return nil, fmt.Errorf("error obtaining end height "+
+			"of file %w", err)
+	}
+
+	// Read the bitcoin network, the headers in the header file belong to.
+	netBytes := [8]byte{}
+	chainChar, err := tlv.ReadVarInt(reader, &netBytes)
```

Comment:
Non-blocking for this iteration, but we can make things more flexible by using a full TLV byte stream here. This way, we can add things to the header over time, without breaking all instances. 

To make room for this in the future though, I think we should add a version byte to the start. Then version `0`, is what we implement in this PR, and we can modify the encoding later. Eg: we can make a compressed version with another version, etc.

Reply:
Okay, so I understand that:
- We can make this a tlv stream
- Add a version byte at the beginning as part of the encoding.
Is that right?

Reply:
Okay so we just read it as it is now then but add version byte